toolchain/check/deferred_definition_worklist.cpp

```diff
@@ -0,0 +1,119 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/deferred_definition_worklist.h"
+
+#include "common/variant_helpers.h"
+#include "common/vlog.h"
+#include "toolchain/check/handle.h"
+
+namespace Carbon::Check {
+
+static constexpr llvm::StringLiteral VlogPrefix = "DeferredDefinitionWorklist ";
+
+DeferredDefinitionWorklist::DeferredDefinitionWorklist(
+    llvm::raw_ostream* vlog_stream)
+    : vlog_stream_(vlog_stream) {
+  // See declaration of `worklist_`.
+  worklist_.reserve(64);
+}
+
+auto DeferredDefinitionWorklist::SuspendFunctionAndPush(
+    Context& context, Parse::DeferredDefinitionIndex index,
+    Parse::FunctionDefinitionStartId node_id) -> void {
+  worklist_.push_back(CheckSkippedDefinition{
+      index, HandleFunctionDefinitionSuspend(context, node_id)});
```

Comment:
This call to `HandleFunctionDefinitionSuspend` is really breaking the abstraction that these two factored out classes are trying to provide -- this isn't a reusable traversal because it also has this weird side-effect of calling a `Handle` function.

Please can you add a `TODO` to move this call out of the worklist?